cirq-core/cirq/transformers/routing/route_circuit_cqc.py

```diff
@@ -0,0 +1,445 @@
+
+            # swaps applied in the current timestep thus far. This ensures the same swaps
+            # don't get executed twice in the same timestep.
+            seen: Set[Tuple[Tuple['cirq.Qid', cirq.Qid], ...]] = set()
+
+            while len(two_qubit_ops[idx]) != 0:
+                chosen_swaps: Optional[Tuple[QidPair, ...]] = None
+                for strat in strats:
+                    chosen_swaps = strat(mm, two_qubit_ops, idx, max_search_radius)
+                    if chosen_swaps is not None:
+                        break
+
+                if chosen_swaps is None or chosen_swaps in seen:
+                    chosen_swaps = cls._brute_force_strategy(mm, two_qubit_ops, idx)
+                else:
+                    seen.add(chosen_swaps)
+
+                for swap in chosen_swaps:
+                    inserted_swap = mm.mapped_op(ops.SWAP(*swap))
+                    if tag_inserted_swaps:
+                        inserted_swap = inserted_swap.with_tags(ops.RoutingSwapTag())
+                    routed_ops[idx].append(inserted_swap)
+                    mm.apply_swap(*swap)
+
+                process_executable_two_qubit_ops(idx)
+
+        return routed_ops
+
+    @classmethod
+    def _brute_force_strategy(
+        cls,
+        mm: mapping_manager.MappingManager,
+        two_qubit_ops: List[List['cirq.Operation']],
+        idx: int,
+    ) -> Tuple[QidPair, ...]:
+        """Inserts SWAPS along the shortest path of the qubits that are the farthest.
+
+        Since swaps along the shortest path are being executed one after the other, in order
+        to achieve the physical swaps (M[q1], M[q2]), (M[q2], M[q3]), ..., (M[q_{i-1}], M[q_i]),
+        we must execute the logical swaps (q1, q2), (q1, q3), ..., (q_1, qi).
+        """
+        furthest_op = max(two_qubit_ops[idx], key=lambda op: mm.dist_on_device(*op.qubits))
+        path = mm.shortest_path(*furthest_op.qubits)
+        return tuple([(path[0], path[i + 1]) for i in range(len(path) - 2)])
+
+    @classmethod
+    def _iterate_swap_pair_strategy(
+        cls,
+        mm: mapping_manager.MappingManager,
+        two_qubit_ops: List[List['cirq.Operation']],
+        idx: int,
+        max_search_radius: int,
+    ) -> Optional[Tuple[QidPair, ...]]:
+        """Computes cost function with pairs of candidate swaps that act on disjoint qubits."""
+        pair_sigma = disjoint_nc2_combinations(cls._initial_candidate_swaps(mm, two_qubit_ops[idx]))
+        return cls._iterate_swaps(mm, two_qubit_ops, idx, max_search_radius, pair_sigma)
+
+    @classmethod
+    def _iterate_single_swap_strategy(
+        cls,
+        mm: mapping_manager.MappingManager,
+        two_qubit_ops: List[List['cirq.Operation']],
+        idx: int,
+        max_search_radius: int,
+    ) -> Optional[Tuple[QidPair, ...]]:
+        """Computes cost function with list of single candidate swaps."""
+        sigma: List[Tuple[QidPair, ...]] = [
```

Comment:
mypy auto infers the type if you directly construct an object. However, if you are initializing an empty list of a certain type, you need to be explicit about the type annotation. 
```suggestion
        sigma = [
```

Reply:
mypy here infers `List[Tuple[QidPair]]` instead of `List[Tuple[QidPair, ...]]` which causes problems elsewhere.